spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/TransformerLayer.scala

```diff
@@ -0,0 +1,366 @@
+      decoderSelfAttentionBias, encoderOutput, attentionBias)
+  }
+
+
+  private[nn] def encodeStack(numLayers: Int,
+                              encoderInput: ModuleNode[T],
+                              attentionBias: ModuleNode[T]): ModuleNode[T] = {
+    decodeStack(numLayers, encoderInput, attentionBias, preName = "encode")
+  }
+
+  private[nn] def decodeStack(numLayers: Int,
+                              decoderInput: ModuleNode[T],
+                              decoderSelfAttentionBias: ModuleNode[T],
+                              encoderOutputs: ModuleNode[T] = null,
+                              attentionBias: ModuleNode[T] = null,
+                              preName: String = "decode"): ModuleNode[T] = {
+    var input = decoderInput
+    var i = 0
+    while (i < numLayers) {
+      val selfAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
+      val selfAttentionModel = prePostProcessingSelfAttention(
+        selfAttention, input, decoderSelfAttentionBias,
+        s"${preName}_self_attention_${i}")
+      input = selfAttentionModel
+
+      if (encoderOutputs != null && attentionBias != null) {
+        val encdecAttention = new Attention[T](hiddenSize, numHeads, attentionDropout)
+        val encdecAttentionModel = prePostProcessingEncDecAttention(
+          encdecAttention, input, encoderOutputs, attentionBias,
+          s"${preName}_encdec_attention_${i}")
+        input = encdecAttentionModel
+      }
+
+      val ffn = new FeedForwardNetwork[T](hiddenSize, filterSize, reluDropout)
+      val ffnModel = prePostProcessingFFN(ffn, input, s"${preName}_ffn_${i}")
+      input = ffnModel
+
+      i += 1
+    }
+    val norm = new LayerNormalization[T](hiddenSize).inputs(input)
+    norm
+  }
+
+  private def prePostProcessingSelfAttention(layer: Module[T], decoderInput: ModuleNode[T],
+    decoderSelfAttentionBias: ModuleNode[T], preName: String): ModuleNode[T] = {
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+        .inputs(decoderInput)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+        .inputs(layer.setName(preName + "/self_attention")
+        .inputs(norm, norm, decoderSelfAttentionBias))
+    CAddTable().inputs(decoderInput, drop)
+  }
+
+  private def prePostProcessingEncDecAttention(
+    layer: Module[T],
+    decoderInput: ModuleNode[T],
+    encoderOutput: ModuleNode[T],
+    attentionBias: ModuleNode[T], preName: String): ModuleNode[T] = {
+    val norm = new LayerNormalization[T](hiddenSize).setName(preName + "/norm")
+      .inputs(decoderInput)
+    val drop = Dropout[T](1 - postprocessDropout).setName(preName + "/dropout")
+      .inputs(layer.setName(preName + "/encdec_attention")
+        .inputs(norm, encoderOutput, attentionBias))
+    CAddTable().inputs(decoderInput, drop)
+  }
+
+  private def prePostProcessingFFN(layer: Module[T],
+    decoderInput: ModuleNode[T], preName: String): ModuleNode[T] = {
```

Comment:
maybe change `preName` to `processType` ？

Reply:
preName is added as part of layer name, e.g. `preName` + `realName` 
so I think `preName` is more proper.